unmarshal.go

```diff
@@ -232,7 +232,14 @@ func (ro *resourceObject) unmarshalFields(v any, rv reflect.Value, rt reflect.Ty
 			if setPrimary {
 				return ErrUnmarshalDuplicatePrimaryField
 			}
-			if ro.Type != jsonapiTag.resourceType {
+
+			if vut, ok := v.(UnmarshalType); ok {
+				err := vut.UnmarshalType(ro.Type)
+				if err != nil {
+					return &TypeError{Actual: ro.Type, Expected: []string{jsonapiTag.resourceType}}
```

Comment:
We are losing the error info that the implementer might have provided in the returned error.
We could:
```
errors.Join(
        &TypeError{Actual: ro.Type, Expected: []string{jsonapiTag.resourceType}},
        err,
)
```